### ISSUE TITLE
ICE after simplification

### ISSUE THREAD

**Issue by Munksgaard** [diku-dk/futhark#1519]
The following IR causes an ICE when passed to `futhark dev -e`:

```futhark

entry("main",
      {points: direct, numBins: direct, numBins2: direct, threshold: direct},
      {direct})
  entry_main (points_mem_4965 : mem, numD_4928 : i64,
              points_4929 : [numD_4928]f32 @ points_mem_4965 ->
                            {base: [numD_4928]; contiguous: true;
                             LMADs: [{offset: 0i64; strides: [1i64];
                                      rotates: [0i64]; shape: [numD_4928];
                                      permutation: [0]; monotonicity: [Inc]}]},
              numBins_4930 : i32, numBins2_4931 : i64, threshold_4932 : f32)
  : {mem, [numD_4928][numBins2_4931]i32 @ ?0->
          {base: [numD_4928, numBins2_4931]; contiguous: true;
           LMADs: [{offset: 0i64; strides: [numBins2_4931, 1i64];
                    rotates: [0i64, 0i64]; shape: [numD_4928, numBins2_4931];
                    permutation: [0, 1]; monotonicity: [Inc, Inc]}]}} = {
  let {binop_x_4966 : i64} = mul_nw64(numD_4928, numBins2_4931)
  let {binop_y_4967 : i64} = mul_nw64(4i64, binop_x_4966)
  let {bytes_4968 : i64} = smax64(0i64, binop_y_4967)
  let {mem_4969 : mem} =
    alloc(bytes_4968)
  let {binop_y_4980 : i64} = mul_nw64(4i64, numBins2_4931)
  let {bytes_4981 : i64} = smax64(0i64, binop_y_4980)
  let {mem_4982 : mem} =
    alloc(bytes_4981)
  let {mem_5020 : mem} =
    alloc(bytes_4968)
  let {result_4952 : [numD_4928][numBins2_4931]i32 @ mem_5020 ->
                     {base: [numD_4928, numBins2_4931]; contiguous: true;
                      LMADs: [{offset: 0i64; strides: [numBins2_4931, 1i64];
                               rotates: [0i64, 0i64];
                               shape: [numD_4928, numBins2_4931];
                               permutation: [0, 1];
                               monotonicity: [Inc, Inc]}]}} =
    scratch(i32, numD_4928, numBins2_4931)
  let {defunc_1_map_res_4935 : [numBins2_4931]i32 @ mem_4982 ->
                               {base: [numBins2_4931]; contiguous: true;
                                LMADs: [{offset: 0i64; strides: [1i64];
                                         rotates: [0i64];
                                         shape: [numBins2_4931];
                                         permutation: [0];
                                         monotonicity: [Inc]}]}} =
    replicate([numBins2_4931], 0i32)
  let {defunc_2_map_res_4933 : [numD_4928][numBins2_4931]i32 @ mem_5020 ->
                               {base: [numD_4928, numBins2_4931];
                                contiguous: true; LMADs: [{offset: 0i64;
                                                           strides: [numBins2_4931, 1i64];
                                                           rotates: [0i64, 0i64];
                                                           shape: [numD_4928, numBins2_4931];
                                                           permutation: [0, 1];
                                                           monotonicity: [Inc, Inc]}]}} =
    loop {mapout_4953 : *[numD_4928][numBins2_4931]i32 @ mem_5020 ->
                        {base: [numD_4928, numBins2_4931]; contiguous: true;
                         LMADs: [{offset: 0i64; strides: [numBins2_4931, 1i64];
                                  rotates: [0i64, 0i64];
                                  shape: [numD_4928, numBins2_4931];
                                  permutation: [0, 1];
                                  monotonicity: [Inc, Inc]}]}} = {result_4952}
    for i_4954:i64 < numD_4928 do {
      let {x_4934 : f32} =
        points_4929[i_4954]
      let {cond_4936 : bool} = lt32(threshold_4932, x_4934)
      let {modified_source_4957 : [numD_4928][numBins2_4931]i32 @ mem_5020 ->
                                  {base: [numD_4928, numBins2_4931];
                                   contiguous: true; LMADs: [{offset: 0i64;
                                                              strides: [numBins2_4931, 1i64];
                                                              rotates: [0i64, 0i64];
                                                              shape: [numD_4928, numBins2_4931];
                                                              permutation: [0, 1];
                                                              monotonicity: [Inc, Inc]}]}} =
        mapout_4953 with [i_4954,
                          0i64 :+ numBins2_4931 * 1i64] = defunc_1_map_res_4935
      let {lw_dest_4955 : [numD_4928][numBins2_4931]i32 @ mem_5020 ->
                          {base: [numD_4928, numBins2_4931]; contiguous: true;
                           LMADs: [{offset: 0i64;
                                    strides: [numBins2_4931, 1i64];
                                    rotates: [0i64, 0i64];
                                    shape: [numD_4928, numBins2_4931];
                                    permutation: [0, 1];
                                    monotonicity: [Inc, Inc]}]}} =
        loop {lowered_array_4956 : *[numD_4928][numBins2_4931]i32 @ mem_5020 ->
                                   {base: [numD_4928, numBins2_4931];
                                    contiguous: true; LMADs: [{offset: 0i64;
                                                               strides: [numBins2_4931, 1i64];
                                                               rotates: [0i64, 0i64];
                                                               shape: [numD_4928, numBins2_4931];
                                                               permutation: [0, 1];
                                                               monotonicity: [Inc, Inc]}]}} = {modified_source_4957}
        for j_4938:i32 < numBins_4930 do {
          let {dBins_row_4958 : [numBins2_4931]i32 @ mem_5020 ->
                                {base: [numD_4928, numBins2_4931];
                                 contiguous: true;
                                 LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                          strides: [1i64]; rotates: [0i64];
                                          shape: [numBins2_4931];
                                          permutation: [0];
                                          monotonicity: [Inc]}]}} =
            lowered_array_4956[i_4954, 0i64 :+ numBins2_4931 * 1i64]
          let {ext_mem_5000 : mem,
               loopres_4940 : [numBins2_4931]i32 @ mem_5020 ->
                              {base: [numD_4928, numBins2_4931];
                               contiguous: true;
                               LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                        strides: [1i64]; rotates: [0i64];
                                        shape: [numBins2_4931];
                                        permutation: [0];
                                        monotonicity: [Inc]}]}} =
            if cond_4936
            then {
              let {mem_4996 : mem} =
                alloc(bytes_4981)
              let {i_4941 : i32} = add32(1i32, numBins_4930)
              let {i_4942 : i64} = sext i32 i_4941 to i64
              let {x_4943 : i32} =
                lowered_array_4956[i_4954, i_4942]
              let {lw_val_4944 : i32} = add32(1i32, x_4943)
              let {dBins_row_copy_4961 : [numBins2_4931]i32 @ mem_5020 ->
                                         {base: [numD_4928, numBins2_4931];
                                          contiguous: true;
                                          LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                                   strides: [1i64];
                                                   rotates: [0i64];
                                                   shape: [numBins2_4931];
                                                   permutation: [0];
                                                   monotonicity: [Inc]}]}} =
                copy(dBins_row_4958)
              let {dBins_4945 : [numBins2_4931]i32 @ mem_5020 ->
                                {base: [numD_4928, numBins2_4931];
                                 contiguous: true;
                                 LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                          strides: [1i64]; rotates: [0i64];
                                          shape: [numBins2_4931];
                                          permutation: [0];
                                          monotonicity: [Inc]}]}} =
                dBins_row_copy_4961 with [i_4942] = lw_val_4944
              let {dBins_4946 : [numBins2_4931]i32 @ mem_5020 ->
                                {base: [numD_4928, numBins2_4931];
                                 contiguous: true;
                                 LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                          strides: [1i64]; rotates: [0i64];
                                          shape: [numBins2_4931];
                                          permutation: [0];
                                          monotonicity: [Inc]}]}} = dBins_4945
              in {mem_4996, dBins_4945}
            } else {
              let {mem_4999 : mem} =
                alloc(bytes_4981)
              let {numBins_4947 : i64} = sext i32 numBins_4930 to i64
              let {x_4948 : i32} =
                lowered_array_4956[i_4954, numBins_4947]
              let {lw_val_4949 : i32} = add32(1i32, x_4948)
              let {dBins_row_copy_4963 : [numBins2_4931]i32 @ mem_5020 ->
                                         {base: [numD_4928, numBins2_4931];
                                          contiguous: true;
                                          LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                                   strides: [1i64];
                                                   rotates: [0i64];
                                                   shape: [numBins2_4931];
                                                   permutation: [0];
                                                   monotonicity: [Inc]}]}} =
                copy(dBins_row_4958)
              let {dBins_4950 : [numBins2_4931]i32 @ mem_5020 ->
                                {base: [numD_4928, numBins2_4931];
                                 contiguous: true;
                                 LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                          strides: [1i64]; rotates: [0i64];
                                          shape: [numBins2_4931];
                                          permutation: [0];
                                          monotonicity: [Inc]}]}} =
                dBins_row_copy_4963 with [numBins_4947] = lw_val_4949
              let {dBins_4951 : [numBins2_4931]i32 @ mem_5020 ->
                                {base: [numD_4928, numBins2_4931];
                                 contiguous: true;
                                 LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                          strides: [1i64]; rotates: [0i64];
                                          shape: [numBins2_4931];
                                          permutation: [0];
                                          monotonicity: [Inc]}]}} = dBins_4950
              in {mem_4999, dBins_4950}
            }
            : {mem, [numBins2_4931]i32 @ (mem_5020 ->
                                          {base: [numD_4928, numBins2_4931];
                                           contiguous: true;
                                           LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                                    strides: [1i64];
                                                    rotates: [0i64];
                                                    shape: [numBins2_4931];
                                                    permutation: [0];
                                                    monotonicity: [Inc]}]})}
          let {lowered_array_updated_4964 : [numD_4928][numBins2_4931]i32 @ mem_5020 ->
                                            {base: [numD_4928, numBins2_4931];
                                             contiguous: true;
                                             LMADs: [{offset: 0i64;
                                                      strides: [numBins2_4931, 1i64];
                                                      rotates: [0i64, 0i64];
                                                      shape: [numD_4928, numBins2_4931];
                                                      permutation: [0, 1];
                                                      monotonicity: [Inc, Inc]}]}} =
            lowered_array_4956 with [i_4954,
                                     0i64 :+ numBins2_4931 * 1i64] = loopres_4940
          in {lowered_array_updated_4964}
        }
      in {lw_dest_4955}
    }
  let {defunc_2_map_res_linear_5021 : [numD_4928][numBins2_4931]i32 @ mem_5020 ->
                                      {base: [numD_4928, numBins2_4931];
                                       contiguous: true; LMADs: [{offset: 0i64;
                                                                  strides: [numBins2_4931, 1i64];
                                                                  rotates: [0i64, 0i64];
                                                                  shape: [numD_4928, numBins2_4931];
                                                                  permutation: [0, 1];
                                                                  monotonicity: [Inc, Inc]}]}} =
    copy(defunc_2_map_res_4933)
  in {mem_5020, defunc_2_map_res_linear_5021}
}
```

The output:

```
$ futhark-master dev -e  issue194.fut_seq_mem
Internal compiler error.  Please report this:
  https://github.com/diku-dk/futhark/issues
Type error after pass 'simplify':
In function entry_main
When checking function body
In expression of statement
  {defunc_2_map_res_4933 : ({}, [numD_4928][numBins2_4931]i32 @ mem_5020 ->
                            {base: [numD_4928, numBins2_4931]; contiguous: true;
                             LMADs: [{offset: 0i64;
                                      strides: [numBins2_4931, 1i64];
                                      rotates: [0i64, 0i64];
                                      shape: [numD_4928, numBins2_4931];
                                      permutation: [0, 1];
                                      monotonicity: [Inc, Inc]}]})}
Inside the loop body
In expression of statement
  {lw_dest_4955 : ({}, [numD_4928][numBins2_4931]i32 @ mem_5020 ->
                   {base: [numD_4928, numBins2_4931]; contiguous: true;
                    LMADs: [{offset: 0i64; strides: [numBins2_4931, 1i64];
                             rotates: [0i64, 0i64];
                             shape: [numD_4928, numBins2_4931];
                             permutation: [0, 1]; monotonicity: [Inc, Inc]}]})}
Inside the loop body
Type error:
Variable lowered_array_4956 referenced after being consumed.
```

The IR has been generated by calling `futhark dev -s -f -e --in-place-lowering -a --cse -e -e --lift-allocations -e --short-circuit tests/issue194.fut` with the latest version from the `old-coalesce` branch (7c2a70bd4).

**Comment by athas** [comment]
Looks like an invalid copy removal.  These used to be hell to fix, but the simplifier improvements I made recently should make it more viable.

**Comment by athas** [comment]
Actually, it turns out that it's the short-circuiting that introduces aliasing at the memory level in an invalid manner!

We have

```
	  let {dBins_row_4958 : [numBins2_4931]i32 @ mem_5020 ->
	                        {base: [numD_4928, numBins2_4931];
				 contiguous: true;
				 LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
				          strides: [1i64]; rotates: [0i64];
				          shape: [numBins2_4931];
				          permutation: [0];
				          monotonicity: [Inc]}]}} =
		lowered_array_4956[i_4954, 0i64 :+ numBins2_4931 * 1i64]
```

So we take a slice of `lowered_array_4956`.  Later there is a copy of that slice that has (presumably) been optimised to be located in exactly the same bit of memory:

```
              let {dBins_row_copy_4961 : [numBins2_4931]i32 @ mem_5020 ->
                                         {base: [numD_4928, numBins2_4931];
                                          contiguous: true;
                                          LMADs: [{offset: mul_nw64 (i_4954) (numBins2_4931);
                                                   strides: [1i64];
                                                   rotates: [0i64];
                                                   shape: [numBins2_4931];
                                                   permutation: [0];
                                                   monotonicity: [Inc]}]}} =
                copy(dBins_row_4958)
```

But that's not really a copy anymore, now is it?  The type checker cannot detect that this is bad (but it is), but there is a simplification rule that removes copies where source and destination is *exactly* the same (in memory terms).  This rule is correct, but it reveals the problem to the type checker.

**Comment by Munksgaard** [comment]
But that's exactly the point of the short-circuiting: to remove copies. In this case, it's just done by making the copy an effective no-op. Shouldn't it just be removed by the simplifier? Or is the problem that the `dBins_row_4958` is not unqiue, and therefore updating it in-place is not correct?

**Comment by athas** [comment]
It is indeed removed by the simplifier, and then you are left with a consumption violation.  You are consuming a slice of an array (`dBins_row_4958`), which also consumes all aliases, and then later you are using the whole array (`lowered_array_4956`).

**Comment by Munksgaard** [comment]
Fixed by f051ab0d8
